src/i2c.rs

```diff
@@ -111,14 +111,20 @@ macro_rules! busy_wait {
 
 impl<I2C, SCL, SDA> I2c<I2C, (SCL, SDA)> {
     /// Configures the I2C peripheral to work in master mode
-    pub fn new<F>(i2c: I2C, pins: (SCL, SDA), freq: F, clocks: Clocks, apb1: &mut APB1) -> Self
+    pub fn new<F>(
+        i2c: I2C,
+        pins: (SCL, SDA),
+        freq: F,
+        clocks: Clocks,
+        apb1: &mut APB1,
+    ) -> Result<Self, <F as TryInto<Hertz<u32>>>::Error>
     where
         I2C: Instance,
         SCL: SclPin<I2C>,
         SDA: SdaPin<I2C>,
-        F: Into<Hertz>,
+        F: Rate + TryInto<Hertz<u32>>,
     {
-        let freq = freq.into().0;
+        let freq = (freq.try_into()? as Hertz).0;
```

Comment:
This is a breaking change and should be noted in the CHANGELOG.md

But I'm not really a fan of this approch, as `F`, which now has to implement `TryInto` is validated inside the function, which in turn just returns the conversion error of `TryInto`. 

I'd suggest, so that the user has to validate the conversion themselves, before passing it into the function:

```suggestion
    pub fn new<F>(
        i2c: I2C,
        pins: (SCL, SDA),
        freq: Hertz,
        clocks: Clocks,
        apb1: &mut APB1,
    ) -> Self
    where
        I2C: Instance,
        SCL: SclPin<I2C>,
        SDA: SdaPin<I2C>,
    {
        let freq = freq.0;
```